metrics/exporter.go

```diff
@@ -32,11 +33,107 @@ type flushable interface {
+	// "activator", "queue_proxy". Should only contains alphabets and underscore.
+	// Must be present.
+	Component string
+
+	// PrometheusPort is the port to expose metrics if metrics backend is Prometheus.
+	// It should be between maxPrometheusPort and maxPrometheusPort. 0 value means
+	// using the default 9090 value. If is ignored if metrics backend is not
+	// Prometheus.
+	PrometheusPort int
+
+	// ConfigMap is the data from config map config-observability. Must be present.
+	// See https://github.com/knative/serving/blob/master/config/config-observability.yaml
+	// for details.
+	ConfigMap map[string]string
+}
+
+// UpdateExporterFromConfigMap returns a helper func that can be used to update the exporter
+// when a config map is updated.
+func UpdateExporterFromConfigMap(component string, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	domain := Domain()
+	return func(configMap *corev1.ConfigMap) {
+		UpdateExporter(ExporterOptions{
+			Domain:    domain,
+			Component: component,
+			ConfigMap: configMap.Data,
+		}, logger)
+	}
+}
+
+// UpdateExporter updates the exporter based on the given ExporterOptions.
+func UpdateExporter(ops ExporterOptions, logger *zap.SugaredLogger) error {
+	metricsMux.Lock()
+	defer metricsMux.Unlock()
+
+	newConfig, err := createMetricsConfig(ops, logger)
+	if err != nil {
+		if curMetricsExporter == nil {
+			// Fail the process if there doesn't exist an exporter.
+			logger.Errorw("Failed to get a valid metrics config", zap.Error(err))
+		} else {
+			logger.Errorw("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+		}
+		return err
+	}
+
+	if isNewExporterRequiredHelper(newConfig, curMetricsConfig) {
+		logger.Info("Flushing the existing exporter before setting up the new exporter.")
+		flushExporterHelper(curMetricsExporter)
+		e, err := newMetricsExporterHelper(newConfig, curMetricsExporter, logger)
+		if err != nil {
+			logger.Errorf("Failed to update a new metrics exporter based on metric config %v. error: %v", newConfig, err)
+			return err
+		}
+		existingConfig := curMetricsConfig
+		setCurMetricsExporterUnlocked(e)
+		logger.Infof("Successfully updated the metrics exporter; old config: %v; new config %v", existingConfig, newConfig)
+	}
+
+	setCurMetricsConfigUnlocked(newConfig)
+	return nil
+}
+
+// isNewExporterRequired compares the non-nil newConfig against curMetricsConfig. When backend changes,
+// or stackdriver project ID changes for stackdriver backend, we need to update the metrics exporter.
+func isNewExporterRequired(newConfig *metricsConfig) bool {
+	cc := getCurMetricsConfig()
+	return isNewExporterRequiredHelper(newConfig, cc)
+}
+
+func isNewExporterRequiredHelper(newConfig *metricsConfig, cc *metricsConfig) bool {
+	if cc == nil || newConfig.backendDestination != cc.backendDestination {
+		return true
+	} else if newConfig.backendDestination == Stackdriver && newConfig.stackdriverProjectID != cc.stackdriverProjectID {
```

Comment:
replace whole else if block with
```
if ... { return true }
return a== b && c != d
```

Or just completely flatten it out into
`return a==b  || c != d || e ==f || g != h`.

Reply:
Pending